libs/ml/tests/ml/layers/fully_connected.cpp

```diff
@@ -80,6 +81,169 @@ TYPED_TEST(FullyConnectedTest, ops_backward_test)  // Use the class as an Ops
+      g_not_shared_statedict_before.dict_["FC4_FullyConnected_Bias"].weights_->Copy());
+  g_not_shared_weights_before.emplace_back(
+      g_not_shared_statedict_before.dict_["FC5_FullyConnected_Weights"].weights_->Copy());
+  g_not_shared_weights_before.emplace_back(
+      g_not_shared_statedict_before.dict_["FC5_FullyConnected_Bias"].weights_->Copy());
+
+  for (size_t i = 0; i < 4; i++)
+  {
+    ASSERT_EQ(g_shared_weights_before[i], g_not_shared_weights_before[i]);
+  }
+
+  // start training
+  // set data
+  ArrayType data;
+  data.Resize({10, 1});
+  for (int i = 0; i < 10; i++)
+  {
+    data.Set(i, 0, DataType(i));
+  }
+
+  // Run 1 iteration of training on g shared
+  auto                                           lr = static_cast<DataType>(1);
+  fetch::ml::optimisers::SGDOptimiser<ArrayType> g_shared_optimiser(
+      g_shared, {g_shared_input}, g_shared_label, g_shared_error, lr);
+  g_shared_optimiser.Run({data}, data, 1);
+  // Run 1 iteration of training on g not shared
+  fetch::ml::optimisers::SGDOptimiser<ArrayType> g_not_shared_optimiser(
+      g_not_shared, {g_not_shared_input}, g_not_shared_label, g_not_shared_error, lr);
+  g_not_shared_optimiser.Run({data}, data, 1);
+
+  // check that all weights are equal
+  std::cout << "**************************AFTER TRAINING**************************" << std::endl;
+  auto                   g_shared_statedict_after = g_shared->StateDict();
+  std::vector<ArrayType> g_shared_weights_after;
+  g_shared_weights_after.emplace_back(
+      g_shared_statedict_after.dict_["FC1_FullyConnected_Weights"].weights_->Copy());
+  g_shared_weights_after.emplace_back(
+      g_shared_statedict_after.dict_["FC1_FullyConnected_Bias"].weights_->Copy());
+  g_shared_weights_after.emplace_back(
+      g_shared_statedict_after.dict_["FC1_Copy_1_FullyConnected_Weights"].weights_->Copy());
+  g_shared_weights_after.emplace_back(
+      g_shared_statedict_after.dict_["FC1_Copy_1_FullyConnected_Bias"].weights_->Copy());
+
+  auto                   g_not_shared_statedict_after = g_not_shared->StateDict();
+  std::vector<ArrayType> g_not_shared_weights_after;
+  g_not_shared_weights_after.emplace_back(
+      g_not_shared_statedict_after.dict_["FC4_FullyConnected_Weights"].weights_->Copy());
+  g_not_shared_weights_after.emplace_back(
+      g_not_shared_statedict_after.dict_["FC4_FullyConnected_Bias"].weights_->Copy());
+  g_not_shared_weights_after.emplace_back(
+      g_not_shared_statedict_after.dict_["FC5_FullyConnected_Weights"].weights_->Copy());
+  g_not_shared_weights_after.emplace_back(
+      g_not_shared_statedict_after.dict_["FC5_FullyConnected_Bias"].weights_->Copy());
+
+  // check the shared weight graph has tied weights, and the not shared weight graph has untied
+  // weights
+  for (size_t i = 0; i < 2; i++)
+  {
+    ASSERT_TRUE(g_shared_weights_before[i] == g_shared_weights_before[i + 2]);
+    ASSERT_TRUE(g_not_shared_weights_before[i] == g_not_shared_weights_before[i + 2]);
```

Comment:
this looks like it checks that the untied weights are actually the same as each other (which is the opposite of what the comment claims"